File: js/2017/01-InverseCaptcha/solution.js
fs = require("fs");
input = fs.readFileSync("input.txt", "ascii");

function solution_part1() {
	solution = 0;
	for (i = 0; i < input.length; i++) {
		const char = input[i];
		const next = input[(i + 1) % input.length]
		if (char == next) {
			solution += Number(char);
		}
	}
	return solution;
}

function solution_part2() {
	solution = 0;
	for (i = 0; i < input.length; i++) {
		const char = input[i];
		const next = input[(i + input.length / 2) % input.length]
		if (char == next) {
			solution += Number(char);
		}
	}
	return solution;
}

console.log("Solution Part 1 is " + solution_part1());
console.log("Solution Part 2 is " + solution_part2());

File: go/2017/01-InverseCaptcha/solution.go
package main

import (
	"fmt"
	"os"
)

func solution_part1(input []byte) int {
	solution := 0
	for i := 0; i < len(input); i++ {
		char := input[i]
		next := input[(i + 1) % len(input)]
		if char == next {
			solution += int(char - 48) // Quick conversion from ASCII encoding to numbers (48..57 map to 0..9)
		}
	}
	return solution
}

func solution_part2(input []byte) int {
	solution := 0
	for i := 0; i < len(input); i++ {
		char := input[i]
		next := input[(i + len(input) / 2) % len(input)]
		if char == next {
			solution += int(char - 48)
		}
	}
	return solution
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err == nil {
		fmt.Println(fmt.Sprintf("Solution Part 1: %d", solution_part1(input)))
		fmt.Println(fmt.Sprintf("Solution Part 2: %d", solution_part2(input)))
	}
}